src/shims/fs.rs

```diff
@@ -257,6 +264,197 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>, // Has type `int`
+        pathname_op: OpTy<'tcx, Tag>, // Has type `const char *`
+        flags_op: OpTy<'tcx, Tag>, // Has type `int`
+        _mask_op: OpTy<'tcx, Tag>, // Has type `unsigned int`
+        statxbuf_op: OpTy<'tcx, Tag> // Has type `struct statx *`
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // If statxbuf is not null, we need to produce a proper `MemPlace` to write the results of
+        // the function to it. This is needed because all the `statx` parameters are provided as
+        // `isize`s instead of having the proper types. In particular `statxbuf` should be a `*mut
+        // statx`. If this wasn't the case, we could use `deref_operand` instead.
+
+        // This long path is required instead of `libc::statx` because the `statx` struct and
+        // function share the same namespace.
+        let statx_ty = this
+            .resolve_path(&["libc", "unix", "linux_like", "linux", "gnu", "statx"])?
+            .ty(*this.tcx);
+        let statxbuf_ty = this.tcx.mk_mut_ptr(statx_ty);
+        let statxbuf_layout = this.layout_of(statxbuf_ty)?;
+        let statxbuf_imm = ImmTy::from_scalar(statxbuf_scalar, statxbuf_layout);
+        let statxbuf_place = this.ref_to_mplace(statxbuf_imm)?;
+
+        let pathname: PathBuf = this.read_os_str_from_c_str(pathname_scalar)?.into();
+        let flags = this.read_scalar(flags_op)?.to_machine_isize(&*this.tcx)? as i32;
+
+        let at_empty_path = this.eval_libc_i32("AT_EMPTY_PATH")?;
+        let at_fdcwd = this.eval_libc_i32("AT_FDCWD")?;
+
+        // Now we need to resolve the path of the target file. This was done following the manual
+        // entry for `statx`
+
+        // If `pathname` is absolute we can ignore `dirfd`.
+        let path = if pathname.is_absolute() {
+            pathname
+        } else {
+            let dirfd = this.read_scalar(dirfd_op)?.to_machine_isize(&*this.tcx)? as i32;
```

Comment:
Another lossy cast.

Reply:
sames as `flags_op`. Should I add a comment about that? maybe we should have a function that fixes the operands for syscalls or something.

Reply:
Helper methods are always good.